tests/autoscaler/autoscaler_test.go

```diff
@@ -0,0 +1,201 @@
+
+	// Scale helloworld deployment to len(workers) + 1 replicas to trigger a scale up.
+	expectedWorkersCount := int32(workersCount + 1)
+	logger.LogCtx(ctx, "level", "debug", fmt.Sprintf("Scaling deployment %s/%s to %d replicas", helloWorldNamespace, helloWorldDeploymentName, expectedWorkersCount))
+	err = scaleDeployment(ctx, tcCtrlClient, expectedWorkersCount)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	logger.LogCtx(ctx, "level", "debug", fmt.Sprintf("Waiting for %d worker nodes to exist", expectedWorkersCount))
+
+	// Wait for nodes to increase by one.
+	o := func() error {
+		workersCount, err := getWorkersCount(ctx, tcCtrlClient)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if int32(workersCount) != expectedWorkersCount {
+			return microerror.Maskf(executionFailedError, "Expecting %d workers, %d found", expectedWorkersCount, workersCount)
+		}
+
+		return nil
+	}
+	b := backoff.NewExponential(backoff.LongMaxWait, 10*time.Second)
+	n := backoff.NewNotifier(logger, ctx)
+	err = backoff.RetryNotify(o, b, n)
+	if err != nil {
+		t.Fatalf("timeout waiting for cluster to scale up: %v", err)
+	}
+
+	// Scale down deployment, wait for one node to get deleted.
+	expectedWorkersCount = expectedWorkersCount - 1
+	logger.LogCtx(ctx, "level", "debug", fmt.Sprintf("Scaling deployment %s/%s to %d replicas", helloWorldNamespace, helloWorldDeploymentName, expectedWorkersCount))
+	err = scaleDeployment(ctx, tcCtrlClient, expectedWorkersCount)
+	if err != nil {
+		t.Fatalf("timeout waiting for cluster to scale down: %v", err)
+	}
+
+	err = backoff.RetryNotify(o, b, n)
+	if err != nil {
+		t.Fatalf("timeout waiting for cluster to scale down: %v", err)
+	}
+
+	_ = tcCtrlClient.Delete(ctx, deployment)
+}
+
+func getWorkersCount(ctx context.Context, ctrlClient client.Client) (int, error) {
+	workers := &corev1.NodeList{}
+	err := ctrlClient.List(ctx, workers, client.MatchingLabels{"kubernetes.io/role": "worker"})
+	if err != nil {
+		return -1, err
+	}
+
+	return len(workers.Items), nil
+}
+
+func scaleDeployment(ctx context.Context, ctrlClient client.Client, expectedWorkersCount int32) error {
+	deployment := &appsv1.Deployment{}
+	err := ctrlClient.Get(ctx, client.ObjectKey{Namespace: helloWorldNamespace, Name: helloWorldDeploymentName}, deployment)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	deployment.Spec.Replicas = &expectedWorkersCount
+
+	err = ctrlClient.Update(ctx, deployment)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func createDeployment(ctx context.Context, ctrlClient client.Client, replicas int32) (*appsv1.Deployment, error) {
+	// Check if deployment exists already.
+	existing := &appsv1.Deployment{}
+	err := ctrlClient.Get(ctx, client.ObjectKey{Name: helloWorldDeploymentName, Namespace: helloWorldNamespace}, existing)
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, microerror.Mask(err)
+	}
+
+	if existing.Name != "" {
```

Comment:
Why would the deployment already exist in e2e cluster? Can this hide a race with another test and produce random results from these two conflicting tests? I'd maybe just drop this check and error out in case creation fails.

Reply:
I did this to avoid a test from failing if run more than once on the same cluster (expecially during test development).
The tests are supposed to be run on a specific cluster only once, so I guess this doesn't hurt.
WDYT?

Reply:
Can we make it the other way around and try to delete the `Deployment` after running the test, even when it fails?

Reply:
My concern was more that `default/helloworld` is very generic name for a deployment and can easily conflict with some other e2e test using `helloworld` app.

For example: What will be the deployment name of this: https://github.com/giantswarm/sonobuoy-plugin/pull/1/files#diff-e4ef1448565059fb453d84d625b054b10e7f49b791efee738f720a8dc81b8327R101-R111

Reply:
so should I assume the deployment is not there (and make the test fail if it is)?
I mean, this was added mainly to make the test be repeatable in an existing cluster and that should happen during development of the test itself.

Reply:
I'd vote for assuming that the deployment is not there. That way we will catch in case there's are conflict with another test using the same deployment name. When developing the test, instrumenting the code with this snippet or doing it manually from command line before test execution is very easy so I don't think that is required for final version.

Final call is yours :slightly_smiling_face:

Reply:
agreed, done